fedimint-client/src/sm/executor.rs

```diff
@@ -180,6 +182,48 @@ where
         Ok(())
     }
 
+    /// Adds a number of state machines to the executor atomically with other DB
+    /// changes is `dbtx`, but without actually starting executing them.
+    ///
+    /// Like [`Self::add_state_machines_dbtx`] but useful for recovering
+    /// modules, where to module itself is not yet available (recovered) for
+    /// the executor.
+    pub async fn add_state_machines_inactive_dbtx(
+        &self,
+        dbtx: &mut DatabaseTransaction<'_>,
+        states: Vec<DynState<GC>>,
+    ) -> AddStateMachinesResult {
+        for state in states {
+            if !self
+                .inner
+                .valid_module_ids
+                .contains(&state.module_instance_id())
+            {
+                return Err(AddStateMachinesError::Other(anyhow!("Unknown module")));
+            }
+
+            let is_active_state = dbtx
+                .get_value(&ActiveStateKey::from_state(state.clone()))
+                .await
+                .is_some();
+            let is_inactive_state = dbtx
+                .get_value(&InactiveStateKey::from_state(state.clone()))
+                .await
+                .is_some();
+
+            if is_active_state || is_inactive_state {
+                return Err(AddStateMachinesError::StateAlreadyExists);
+            }
+
+            dbtx.insert_entry(
+                &ActiveStateKey::from_state(state.clone()),
+                &ActiveState::new(),
+            )
+            .await;
+        }
+
+        Ok(())
+    }
```

Comment:
The naming of this fn seems weird. It adds a state as active to the executor without checking if it actually is active. This also means afaik that we *have to* only call it with active states, otherwise inactive ones will get stuck in the active table forever, as they don't have any state transitions by definition.

Either we make the `is_terminal` check not require the actual module context (that seems to be the problem here) or we have to add a filter in the executor that will remove inactive states from the active states table if they end up there by accident.

Reply:
> Either we make the `is_terminal` check not require the actual module context (that seems to be the problem here) or we have to add a filter in the executor that will remove inactive states from the active states table if they end up there by accident.

Detecting terminal ones and handling them correctly seems to simplify the caller and be more robust.

Reply:
> Detecting terminal ones and handling them correctly seems to simplify the caller and generally more robust.

Can we do this in this PR please? I want to avoid forgetting about it after this one is used.

Reply:
🙄 Ok, ok. See the last commit. I wasn't all that confident about how I handled it there, but I hope it's OK.